File: guests/go/scrollto/scrollto.go
// The scroll-to scene (tools/scenes/scrollto.steps): the app scrolls a
// list of messages to a row by key (docs/scroll-to-plan.md), opening at
// the newest one before the first layout, jumping to one on a click,
// staying put on a key no row holds, and following its own send.
package scrollto

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

//go:generate go run dev.kaya/cmd/kaya-gen -type Message -key string
type Message struct {
	Text string
}

func App() *kaya.App {
	app := kaya.NewApp()

	app.Build(func(tx *kaya.Tx) {
		messages := MessageCollection(tx)
		count := tx.Signal("60 messages")
		var list kaya.Widget
		sent := 60

		tx.Mount(tx.Column(func() {
			tx.Label(count).A11yID("count")
			tx.Row(func() {
				tx.Button("jump", func(tx *kaya.Tx) {
					tx.ScrollToRow(list, "m10")
				}).A11yID("jump")
				tx.Button("nowhere", func(tx *kaya.Tx) {
					tx.ScrollToRow(list, "m999")
				}).A11yID("nowhere")
				tx.Button("send", func(tx *kaya.Tx) {
					sent++
					key := fmt.Sprintf("m%d", sent)
					messages.Insert(tx, key, Message{Text: fmt.Sprintf("message %d", sent)})
					tx.Write(count, fmt.Sprintf("%d messages", sent))
					tx.ScrollToRow(list, key)
				}).A11yID("send")
			})
			tx.Scroll(func() {
				// The For's own container is what a scroll_to_row
				// addresses (docs/scroll-to-plan.md S1).
				rows := MessageRows(tx, messages)
				list = rows.Widget()
				for row := range rows.All() {
					row.Label(row.Text())
				}
				tx.SetA11yID(list, "messages")
			}).Grow(1)
		}))

		for i := 1; i <= 60; i++ {
			messages.Insert(tx, fmt.Sprintf("m%d", i), Message{Text: fmt.Sprintf("message %d", i)})
		}
		tx.ScrollToRow(list, "m60")
	})

	return app
}

File: guests/java/dev/kaya/guests/ScrollTo.java
package dev.kaya.guests;

import dev.kaya.KayaApp;
import dev.kaya.KayaGen;

/**
 * The scroll-to scene from the JVM — guests/rust/scrollto.rs,
 * tools/scenes/scrollto.steps. The app scrolls a list of messages to a
 * row by key (docs/scroll-to-plan.md), opening at the newest one before
 * the first layout, jumping to one on a click, staying put on a key no
 * row holds, and following its own send.
 */
public final class ScrollTo {
    @KayaGen(key = "String")
    record Message(String text) {}

    public static void app() {
        KayaApp app = new KayaApp();
        int[] sent = {60};
        KayaApp.Widget[] list = new KayaApp.Widget[1];

        app.build(tx -> {
            var messages = MessageKaya.collection(tx);
            KayaApp.Signal<String> count = tx.signal("60 messages");

            tx.mount(tx.column(col -> {
                tx.label(count).a11yId("count"); // label#0
                tx.row(bar -> {
                    tx.button("jump", t -> t.scrollToRow(list[0], "m10"))
                            .a11yId("jump"); // button#0
                    tx.button("nowhere", t -> t.scrollToRow(list[0], "m999"))
                            .a11yId("nowhere"); // button#1
                    tx.button("send", t -> {
                        sent[0]++;
                        messages.insert(t, "m" + sent[0],
                                new Message("message " + sent[0]));
                        t.write(count, sent[0] + " messages");
                        t.scrollToRow(list[0], "m" + sent[0]);
                    }).a11yId("send"); // button#2
                });
                tx.scroll(box -> { // scroll#0
                    // The For's own container, which scrollToRow
                    // addresses (docs/scroll-to-plan.md S1).
                    var rows = MessageKaya.rows(tx, messages);
                    list[0] = rows.handle;
                    for (var row : rows) {
                        row.label(row.text);
                    }
                    list[0].a11yId("messages");
                }).grow(1);
            }));

            for (int i = 1; i <= 60; i++) {
                messages.insert(tx, "m" + i, new Message("message " + i));
            }
            tx.scrollToRow(list[0], "m60");
            return null;
        });

        app.dispatchLoop();
    }

    private ScrollTo() {}
}
